libgalois/src/analytics/louvain_clustering/louvain_clustering.cpp

```diff
@@ -0,0 +1,631 @@
+/*
+ * This file belongs to the Galois project, a C++ library for exploiting
+ * parallelism. The code is being released under the terms of the 3-Clause BSD
+ * License (a copy is located in LICENSE.txt at the top-level directory).
+ *
+ * Copyright (C) 2018, The University of Texas at Austin. All rights reserved.
+ * UNIVERSITY EXPRESSLY DISCLAIMS ANY AND ALL WARRANTIES CONCERNING THIS
+ * SOFTWARE AND DOCUMENTATION, INCLUDING ANY WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR ANY PARTICULAR PURPOSE, NON-INFRINGEMENT AND WARRANTIES OF
+ * PERFORMANCE, AND ANY WARRANTY THAT MIGHT OTHERWISE ARISE FROM COURSE OF
+ * DEALING OR USAGE OF TRADE.  NO WARRANTY IS EITHER EXPRESS OR IMPLIED WITH
+ * RESPECT TO THE USE OF THE SOFTWARE OR DOCUMENTATION. Under no circumstances
+ * shall University be liable for incidental, special, indirect, direct or
+ * consequential damages or loss of profits, interruption of business, or
+ * related expenses which may arise from use of Software or Documentation,
+ * including but not limited to those resulting from defects in Software and/or
+ * Documentation, or loss or inaccuracy of data of any kind.
+ */
+
+#include "katana/analytics/louvain_clustering/louvain_clustering.h"
+
+#include <deque>
+#include <type_traits>
+
+#include "katana/TypedPropertyGraph.h"
+#include "katana/analytics/ClusteringImplementationBase.h"
+
+using namespace katana::analytics;
+namespace {
+
+template <typename EdgeWeightType>
+struct LouvainClusteringImplementation
+    : public katana::analytics::ClusteringImplementationBase<
+          katana::TypedPropertyGraph<
+              std::tuple<
+                  PreviousCommunityId, CurrentCommunityId,
+                  DegreeWeight<EdgeWeightType>>,
+              std::tuple<EdgeWeight<EdgeWeightType>>>,
+          EdgeWeightType, CommunityType<EdgeWeightType>> {
+  using NodeData = std::tuple<
+      PreviousCommunityId, CurrentCommunityId, DegreeWeight<EdgeWeightType>>;
+  using EdgeData = std::tuple<EdgeWeight<EdgeWeightType>>;
+  using CommTy = CommunityType<EdgeWeightType>;
+  using CommArray = katana::LargeArray<CommTy>;
+
+  using Graph = katana::TypedPropertyGraph<NodeData, EdgeData>;
+  using GNode = typename Graph::Node;
+
+  using Base = katana::analytics::ClusteringImplementationBase<
+      Graph, EdgeWeightType, CommTy>;
+
+  katana::Result<double> LouvainWithoutLockingDoAll(
+      katana::PropertyGraph* pfg, double lower,
+      double modularity_threshold_per_round, uint32_t& iter) {
+    katana::StatTimer TimerClusteringTotal("Timer_Clustering_Total");
+    TimerClusteringTotal.start();
+
+    auto graph_result = Graph::Make(pfg);
+    if (!graph_result) {
+      return graph_result.error();
+    }
+    Graph graph = graph_result.value();
+
+    CommArray c_info;    // Community info
```

Comment:
probably will need to update this `CommArray` too, if you change it in the header file